packer: Deal gracefully with missing manifest keys

ptomato opened the pull request:
If writing a manifest by hand, it's inconvenient to have to list a bunch
of keys with empty array values. Instead, make these keys optional.

https://phabricator.endlessm.com/T16018

ptomato committed:
packer: Deal gracefully with missing manifest keys

If writing a manifest by hand, it's inconvenient to have to list a bunch
of keys with empty array values. Instead, make these keys optional.

https://phabricator.endlessm.com/T16018

tchx84 commented on src/packer.js:
```diff
@@ -114,6 +114,11 @@ var Packer = new Lang.Class ({
+            metadata['tags'].push('EknSetObject');
+        else
+            metadata['tags'] = ['EknSetObject'];
+
```
Not sure this is necessary, I mean, the tools need to make sure to add the "right" tags, and they do:

* https://github.com/endlessm/basin/blob/master/tools/incubator.js#L180
* https://github.com/endlessm/basin/blob/master/tools/processor.js#L99

ptomato replied:
It's convenient though, as otherwise in a manually crafted `sets.json` you have to add `"tags": ["EknSetObject"]` ten times, whereas it could be implicitly understood by virtue of being found under a key called `"sets"` :smile:

tchx84 replied:
The problem is I actually want to move away from hard-coding e-k-l-specific object models into the core bits of basin (to make it more flexible, for other use cases like recipes). Doing this goes in the opposite direction, that is why I suggested that maybe we need a tool for crating permanent shards for e-k-l sets (would include this, and maybe even downloading of the thumbnails).

tchx84 replied:
Following up with this, I definitely need to re-design the basin manifest format itself to make the packer less e-k-l-centrist, so ideally we would not even have a "sets" section anymore, just objects.

If we add this hard-coded ekn specific tag here, I will need you to remove it from the tools so we don't end up adding duplicated tags, like:

```
{
  "childTags": [
    "Sports"
  ],
  "tags": [
    "EknSetObject",
    "EknSetObject"
  ],
  "title": "Sports",
  "featured": false,
  "@id": "ekn:///f99aecef3d12e02dcbb6260bbdd35189c89e6e73",
  "@type": "ekn://_vocab/SetObject"
}
```
And that will indeed move the code farther from what I am envisioning. So, to me it clear that we shouldn't do this, and that we should a) add the tag in the example you wrote for the docs, or b) add another tool that handles this. But a) sounds perfectly reasonable to me.

tchx84 replied:
A simpler middle ground could be to just check if its already included, but acknowledging that will this "feature" of adding the tag automatically in the packer will go away.

tchx84 commented on tools/incubator.js:
```diff
@@ -269,20 +272,29 @@ const Incubator = new Lang.Class({
         fail_with_message(USAGE);
 
-   let incubator = new Incubator(hatch_dir, manifest_path);
+   let incubator = new Incubator(hatch_dir, manifest_path, create_sets);
```
We could move this to the hatch_sets.json file... adding a new property, like `ignore-sets: true`?

tchx84 reviewed (changes requested):
Left two comments

ptomato committed:
incubator: Add ignore-unlisted key to hatch_sets.json

`"ignore-unlisted": true` will not automatically create any sets from
tags, only creating the ones explicitly listed in the JSON file.

https://phabricator.endlessm.com/T16018

tchx84 commented on tools/incubator.js:
```diff
@@ -15,6 +15,8 @@ const Incubator = new Lang.Class({
         this._basin_manifest_path = basin_manifest_path;
         this._manifest = this._read_json_file(GLib.build_filenamev([this._hatch_dir, 'hatch_manifest.json']));
         this._sets = this._read_json_file(GLib.build_filenamev([this._hatch_dir, 'hatch_sets.json']));
+        if (!('tags' in this._sets))
```
You need to ensure that `this._sets` is not `null`, in case the file doesn't exist.

tchx84 reviewed (changes requested):
```
(gjs:7): Gjs-WARNING **: 15:44:43.763: JS ERROR: TypeError: this._sets is null
Incubator<._import_set@/usr/share/basin/tools/incubator.js:179:1
wrapper@resource:///org/gnome/gjs/modules/_legacy.js:82:22
Incubator<._import_hatch/<@/usr/share/basin/tools/incubator.js:235:32
Incubator<._import_hatch@/usr/share/basin/tools/incubator.js:234:9
wrapper@resource:///org/gnome/gjs/modules/_legacy.js:82:22
Incubator<.run@/usr/share/basin/tools/incubator.js:274:24
wrapper@resource:///org/gnome/gjs/modules/_legacy.js:82:22
main@/usr/share/basin/tools/incubator.js:293:4
@<command line>:3:1

Script <command line> threw an exception
```

There is still one missing check, it seems.
